src/pool/cli.c

```diff
@@ -709,6 +711,140 @@ struct pool_connect_arg {
+ * This forces underlying connection to be set up.
+ */
+static void
+warmup(struct dc_pool *pool)
+{
+	static bool         parsed;
+	static bool         enabled = false;
+	crt_context_t       ctx;
+	int                 i;
+	int                 shift;
+	struct pool_target *tgts;
+	sem_t               sem;
+	crt_bulk_t          bulk_hdl;
+	void               *bulk_buf;
+	int                 bulk_len = 4096;
+	d_sg_list_t         sgl;
+	d_iov_t             iov;
+	int                 nr;
+	int                 rc = 0;
+
+	if (parsed && !enabled)
+		/** fast path when disabled */
+		return;
+
+	D_MUTEX_LOCK(&warmup_lock);
+	if (!parsed) {
+		d_getenv_bool("D_POOL_WARMUP", &enabled);
+		parsed = true;
+	}
+	if (!enabled) {
+		D_MUTEX_UNLOCK(&warmup_lock);
+		return;
+	}
+
+	D_ALLOC(bulk_buf, bulk_len);
+	if (bulk_buf == NULL) {
+		D_ERROR("Failed to alloc mem\n");
+		D_MUTEX_UNLOCK(&warmup_lock);
+		return;
+	}
+
+	ctx = daos_get_crt_ctx();
+
+	d_iov_set(&iov, bulk_buf, bulk_len);
+	sgl.sg_nr     = 1;
+	sgl.sg_nr_out = 0;
+	sgl.sg_iovs   = &iov;
+	rc            = crt_bulk_create(ctx, &sgl, CRT_BULK_RW, &bulk_hdl);
+	if (rc < 0) {
+		D_ERROR("Failed to create bulk handle\n");
+		D_MUTEX_UNLOCK(&warmup_lock);
+		return;
+	}
+
+	rc = sem_init(&sem, 0, 0);
+	if (rc < 0) {
+		D_ERROR("Failed to initialize semaphore\n");
+		D_MUTEX_UNLOCK(&warmup_lock);
+		return;
+	}
+
+	/** retrieve all targets from the pool map */
+	nr = pool_map_find_target(pool->dp_map, PO_COMP_ID_ALL, &tgts);
+
+	/** Randomize start order to minimize load for large-scale job */
+	shift = rand() + (int)getpid();
+	if (shift < 0)
+		shift = rand();
+
+	D_DEBUG(DB_TRACE, "Pinging %d targets, shifting at %d\n", nr, shift);
+
+	for (i = 0; i < nr; i++) {
+		crt_endpoint_t             ep;
+		crt_rpc_t                 *rpc = NULL;
+		struct pool_tgt_warmup_in *rpc_in;
+		int                        idx;
+		crt_opcode_t               opcode;
+
+		idx        = (i + shift) % nr;
+		ep.ep_grp  = pool->dp_sys->sy_group;
+		ep.ep_rank = tgts[idx].ta_comp.co_rank;
+		ep.ep_tag  = daos_rpc_tag(DAOS_REQ_TGT, tgts[idx].ta_comp.co_index);
+		opcode     = DAOS_RPC_OPCODE(POOL_TGT_WARMUP, DAOS_POOL_MODULE,
+                                         dc_pool_proto_version ? dc_pool_proto_version
+								   : DAOS_POOL_VERSION);
+		rc         = crt_req_create(ctx, &ep, opcode, &rpc);
+		if (rc != 0) {
+			D_ERROR("Failed to allocate req " DF_RC "\n", DP_RC(rc));
+			goto exit;
+		}
+		D_ASSERTF(rc == 0, "crt_req_create failed; rc=%d\n", rc);
+		rpc_in          = crt_req_get(rpc);
+		rpc_in->tw_bulk = bulk_hdl;
+
+		rc = crt_req_send(rpc, warmup_cb, &sem);
+		if (rc != 0) {
+			D_ERROR("Failed to ping rank=%d:%d, " DF_RC "\n", ep.ep_rank, ep.ep_tag,
+				DP_RC(rc));
+			goto exit;
+		}
+
+		while (sem_trywait(&sem) == -1) {
+			rc = crt_progress(ctx, 0);
+			if (rc && rc != -DER_TIMEDOUT) {
+				D_ERROR("failed to progress context, " DF_RC "\n", DP_RC(rc));
+				break;
+			}
```

Comment:
I'd rather not do that because on systems where this is likely going to be enabled, we will have a lot of client processes and server targets, this will create an RPC storm that might cause timeouts etc.

Reply:
or you could eventually create up to N crt requests with N something like 16 and loop over that until you reach `nr` the number of targets (even if in theory cart should handle that through the quota stuff it would be better to control that at this level I think).

Reply:
well that quota stuff is per process.. so it doesn't control anything if you have thousands or tens of thousands of ranks doing the warmup. anyway i don't disagree, but for the use case currently, this is sufficient and we are going to have more than enough inflight when looking at clients ranks to server rank ratio.